File: librarys/med-common-business/ModuleIM/LibIMBusiness/src/main/java/net/medlinker/imbusiness/ui/repository/ChatSessionRepository.java
package net.medlinker.imbusiness.ui.repository;

import android.util.Base64;

import androidx.lifecycle.MutableLiveData;

import com.medlinker.network.retrofit.error.ErrorConsumer;

import net.medlinker.base.network.HttpResultFunc;
import net.medlinker.im.im.ImMsgRecivedManager;
import net.medlinker.im.realm.MsgSessionDbEntity;
import net.medlinker.im.util.SchedulersCompat;
import net.medlinker.imbusiness.entity.ListStringEntity;
import net.medlinker.imbusiness.network.ApiManager;
import net.medlinker.libhttp.BaseEntity;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Predicate;
import io.realm.Realm;

/**
 * @author hmy
 * @time 2020/9/24 19:43
 */
public class ChatSessionRepository {

    private Disposable mGetHistoryMsgDisposable;

    public boolean getHistoryMsg(MsgSessionDbEntity mSession, final int limit,
                                 final MutableLiveData<ListStringEntity> liveData) {
        if (mGetHistoryMsgDisposable != null && !mGetHistoryMsgDisposable.isDisposed()
                || null == mSession || !mSession.isValid()) {
            liveData.setValue(null);
            return false;
        }
        Observable<BaseEntity<ListStringEntity>> observable = mSession.isGroup()
                ? ApiManager.getMsgApi().getGroupHistoryMsg(mSession.getFromGroup().getId(), 1, limit,
                mSession.getStart())
                : ApiManager.getMsgApi().getSingleChatHistroyMsg(mSession.getFromUser().getId(), 1,
                mSession.getFromUser().getRefrence(), 15, mSession.getStart());
        mGetHistoryMsgDisposable = observable
                .compose(SchedulersCompat.<BaseEntity<ListStringEntity>>applyIoSchedulers())
                .map(new HttpResultFunc<ListStringEntity>())
                .filter(new Predicate<ListStringEntity>() {
                    @Override
                    public boolean test(ListStringEntity listStringEntity) throws Exception {
                        return listStringEntity != null && listStringEntity.getList() != null;
                    }
                })
                .doOnNext(new Consumer<ListStringEntity>() {
                    @Override
                    public void accept(final ListStringEntity listStringEntity) throws Exception {
                        try (Realm realm = Realm.getDefaultInstance()) {
                            realm.executeTransaction(new Realm.Transaction() {
                                @Override
                                public void execute(Realm realm) {
                                    for (String s : listStringEntity.getList()) {
                                        if (s == null) {
                                            continue;
                                        }
                                        byte[] decode = Base64.decode(s, Base64.NO_WRAP);
                                        ImMsgRecivedManager.INSTANCE.saveHistoryMsg(decode, realm);
                                    }
                                }
                            });
                        }
                    }
                })
                .subscribe(new Consumer<ListStringEntity>() {
                    @Override
                    public void accept(final ListStringEntity s) throws Exception {
                        if (null != liveData) {
                            liveData.setValue(s);
                        }
                    }
                }, new ErrorConsumer() {
                    @Override
                    public void accept(Throwable throwable) {
                        super.accept(throwable);
                        if (null != liveData) {
                            liveData.setValue(null);
                        }
                    }
                });
        return true;
    }

}

File: modules/IMBusiness/src/main/java/net/medlinker/imbusiness/ui/repository/ChatSessionRepository.kt
package net.medlinker.imbusiness.ui.repository

import android.annotation.SuppressLint
import android.util.Base64
import androidx.lifecycle.MutableLiveData
import com.medlinker.baseapp.entity.DataListResponse
import com.medlinker.network.retrofit.error.ErrorConsumer
import io.reactivex.disposables.Disposable
import io.realm.Realm
import net.medlinker.base.mvvm.BaseViewModel
import net.medlinker.base.mvvm.VMNetCallback
import net.medlinker.base.mvvm.extend.startRequest
import net.medlinker.base.network.HttpResultFunc
import net.medlinker.im.im.ImMsgRecivedManager
import net.medlinker.im.realm.MsgSessionDbEntity
import net.medlinker.im.util.SchedulersCompat
import net.medlinker.imbusiness.entity.ClinicMemberEntity
import net.medlinker.imbusiness.entity.ImButtonEntity
import net.medlinker.imbusiness.entity.ListStringEntity
import net.medlinker.imbusiness.network.Api
import net.medlinker.imbusiness.network.ApiManager
import net.medlinker.libhttp.host.HostManager
import java.util.*


/**
 * @author hmy
 * @time 12/6/21 13:45
 */
class ChatSessionRepository {

    fun getImButtonList(
        doctorId: Long,
        liveData: MutableLiveData<ArrayList<ImButtonEntity>>?,
        vm: BaseViewModel?
    ) {
        HostManager.getApi(Api::class.java)
            .getImButtons(doctorId)
            .startRequest(object : VMNetCallback<DataListResponse<ImButtonEntity>>(vm) {
                override fun onSuccess(data: DataListResponse<ImButtonEntity>) {
                    liveData?.value = data.list as ArrayList<ImButtonEntity>?
                }

            })
    }

    @SuppressLint("CheckResult")
    fun getUserList(groupId: Long, liveData: MutableLiveData<ClinicMemberEntity>?) {
        ApiManager.getClinicApi().getUserList(groupId)
            .compose(SchedulersCompat.applyIoSchedulers())
            .map(HttpResultFunc())
            .subscribe({
                liveData?.value = it
            }, ErrorConsumer())
    }

    private var mGetHistoryMsgDisposable: Disposable? = null

    fun getHistoryMsg(
        mSession: MsgSessionDbEntity?, limit: Int,
        liveData: MutableLiveData<ListStringEntity?>?
    ): Boolean {
        if (mGetHistoryMsgDisposable != null && !mGetHistoryMsgDisposable!!.isDisposed || null == mSession || !mSession.isValid
        ) {
            liveData!!.value = null
            return false
        }
        val observable = if (mSession.isGroup) ApiManager.getMsgApi().getGroupHistoryMsg(
            mSession.fromGroup.id, 1, limit,
            mSession.start
        ) else ApiManager.getMsgApi().getSingleChatHistroyMsg(
            mSession.fromUser.id, 1,
            mSession.fromUser.refrence.toLong(), 15, mSession.start
        )
        mGetHistoryMsgDisposable = observable
            .compose(SchedulersCompat.applyIoSchedulers())
            .map(HttpResultFunc())
            .filter { listStringEntity -> listStringEntity.list != null }
            .doOnNext { listStringEntity ->
                Realm.getDefaultInstance().use { realm ->
                    realm.executeTransaction { realm ->
                        for (s in listStringEntity.list) {
                            if (s == null) {
                                continue
                            }
                            val decode =
                                Base64.decode(s, Base64.NO_WRAP)
                            ImMsgRecivedManager.INSTANCE.saveHistoryMsg(decode, realm)
                        }
                    }
                }
            }
            .subscribe({ s -> liveData?.setValue(s) }, object : ErrorConsumer() {
                override fun accept(throwable: Throwable) {
                    super.accept(throwable)
                    liveData?.value = null
                }
            })
        return true
    }
}